### PR TITLE
Fixed code generation for unions with fields whose types are the same after type erasure

### PR DESCRIPTION
**Pull request by tomdz**

### CODE REVIEW

File: swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java [hunk 1]
```diff
@@ -42,6 +42,7 @@ public void addField(final FieldContext field)
         this.fields.add(field);
     }
 
+    @Deprecated
```
Comment (ryantimwilson) [inline swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java]: You can probably get rid of this function altogether

File: swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java [hunk 2]
```diff
@@ -51,6 +52,26 @@ public boolean getHasUniqueFieldTypes()
       return fieldTypesSet.size() == fieldTypesList.size();
     }
 
+    public boolean getHasUniqueFieldTypesUnderErasure()
+    {
+        List<String> fieldTypesList = Lists.transform(this.fields, new Function<FieldContext, String>() {
+            public String apply(FieldContext field) {
+                // remove generics
+                String javaType = field.getJavaType();
+                int startTypeParameter = javaType.indexOf('<');
```
Comment (ryantimwilson) [inline swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java]: I wonder if < is a valid character in a Java class (not that people should be using that anyways).

**Comment by ryantimwilson** [comment]
LGTM other than minor comments

File: swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java (hunk 2 above)
Comment (tomdz) [inline swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java]: See https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierStart(char) and https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierPart(char). Tl;dr: no :)

File: swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java (hunk 1 above)
Comment (tomdz) [inline swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java]: I don't know how the swift project handles removal of publicly exposed methods, but other OSS projects usually deprecate methods for one or two releases and then remove them.

Comment (ryantimwilson) [inline swift-generator/src/main/java/com/facebook/swift/generator/template/StructContext.java]: Sounds good, we can deprecate it!

**Comment by SergeyMakarenko** [comment]
I spent some time learning how swift code generation works and I (almost) understand what is going on here. Given that you have passing test for your specific problem we should proceed with this.

**Comment by tomdz** [comment]
Yeah, stringtemplate can be a little complicated to understand :)

**Comment by tomdz** [comment]
Could somebody merge this (I don't have write access) and cut a new release ?

**Pull request merged**